File: pbbh.py
class heap:
	class __node__:
		def __init__(self,key,l,r,p,bfr,afr):
			self.key = key
			self.l = l
			self.r = r
			self.p = p
			self.bfr = bfr
			self.afr = afr
	def __init__(self):
		self.n = 0
		self.root = None
		self.last = None
		self.lastparent = None
	def push(self,key):
		if self.n == 0:
			self.root = [self.__node__(key,None,None,None,None,None)]
			self.last = self.root
			self.lastparent = self.root
		else:
			v = [self.__node__(key,None,None,None,self.last,None)]
			self.last[0].afr = v
			self.last = v
			if self.lastparent[0].r:
				self.lastparent = self.lastparent[0].afr
				self.lastparent[0].l = v
			elif self.lastparent[0].l:
				self.lastparent[0].r = v
			else:
				self.lastparent[0].l = v
			v[0].p = self.lastparent
			while v[0].p:
				if v[0].key < v[0].p[0].key:
					t = v[0].key
					v[0].key = v[0].p[0].key
					v[0].p[0].key = t
				else:
					break
				v = v[0].p
		self.n += 1
	def pop(self):
		if self.n == 1:
			self.root = None
			self.last = None
			self.lastparent = None
			self.n -= 1
		elif self.n:
			temp = self.last
			if self.lastparent[0].r:
				self.lastparent[0].r = None
			else:
				self.lastparent[0].l = None
				if self.n > 3:
					self.lastparent = self.lastparent[0].bfr
			self.root[0].key = temp[0].key
			self.last = self.last[0].bfr
			self.last[0].afr = None
			del temp
			temp = self.root
			while temp[0].l:
				s = temp[0].l
				if temp[0].r:
					if temp[0].r[0].key < s[0].key:
						s = temp[0].r
				if s[0].key < temp[0].key:
					t = s[0].key
					s[0].key = temp[0].key
					temp[0].key = t
				else:
					break
				temp = s
			self.n -= 1
	def top(self):
		return self.root[0].key
	def size(self):
		return self.n
	def empty(self):
		return self.n==0
	def __free__(self,v):
		if v:
			self.__free__(v[0].l)
			self.__free__(v[0].r)
			del v
	def __del__(self):
		self.__free__(self.root)

File: pbbh.cpp
template<class type>
class heap {
	private:
		int n = 0;
		struct node {
			type key;
			node *l, *r, *p, *bfr, *afr;
		};
		node *root = NULL, *last = NULL, *lastparent = NULL;
		void free(node *v) {
			if(v) {
				free(v->l);
				free(v->r);
				delete v;
			}
		}
	public:
		void push(type key) {
			if(n == 0) {
				root = new node{key,NULL,NULL,NULL,NULL,NULL};
				last = root;
				lastparent = root;
			} else {
				node *v = new node{key,NULL,NULL,NULL,last,NULL};
				last->afr = v;
				last = v;
				if(lastparent->r) {
					lastparent = lastparent->afr;
					lastparent->l = v;
				} else if(lastparent->l)
					lastparent->r = v;
				else
					lastparent->l = v;
				v->p = lastparent;
				while(v->p) {
					if(v->key < v->p->key)
						swap(v->key,v->p->key);
					else
						break;
					v = v->p;
				}
			}
			n++;
		}
		void pop() {
			if(n == 1) {
				delete root;
				root = NULL;
				last = NULL;
				lastparent = NULL;
				n--;
			} else if(n) {
				node *temp = last, *s;
				if(lastparent->r)
					lastparent->r = NULL;
				else {
					lastparent->l = NULL;
					if(n > 3)
						lastparent = lastparent->bfr;
				}
				root->key = temp->key;
				last = last->bfr;
				last->afr = NULL;
				delete temp;
				temp = root;
				while(temp->l) {
					s = temp->l;
					if(temp->r)
						if(temp->r->key < s->key)
							s = temp->r;
					if(s->key < temp->key)
						swap(s->key,temp->key);
					else
						break;
					temp = s;
				}
				n--;			
			}
		}
		type top() {
			if(root)
				return root->key;
		}
		int size() {
			return n;
		}
		bool empty() {
			return n == 0;
		}
		~heap() {
			free(root);
		}
};
